web/__init__.py

```diff
@@ -34,30 +63,63 @@
 
 @app.before_first_request
 def startup():
+    """
+    Registers required services on server start
+    """
     app.svc.register("pppp", web.service.pppp.PPPPService())
     app.svc.register("videoqueue", web.service.video.VideoQueue())
     app.svc.register("mqttqueue", web.service.mqtt.MqttQueue())
     app.svc.register("filetransfer", web.service.filetransfer.FileTransferService())
 
 
+def shutdown():
+    """
+    Unregisters not required services on server shutdown
+    """
+    app.svc.unregister("pppp")
+    app.svc.unregister("videoqueue")
+    app.svc.unregister("mqttqueue")
+    app.svc.unregister("filetransfer")
+
+
+def restart():
```

Comment:
This *does not work* correctly.

It may look like it does, and it might occasionally do the right thing, but it will absolutely lead to tricky bugs, and incorrect behavior.

The service manager (`app.svc`) keeps track of a number of registered services (those added with `register()`), and makes them available for use.

The web server (either endpoints, or registered services) can request access to specific services. This is tracked using reference counting.

For example, in this fragment:

```python
...
    for msg in app.svc.stream("videoqueue"):
...
```

The `.stream()` method iterates over messages from a service, and uses the service while doing so. So while streaming, the `videoqueue` service will have a reference count of at least 1.

If we `unregister()` the services, the `ServiceManager` removes the service registration, as well as the reference count to them. However, the actual `Service` objects will still be allocated, causing resource conflicts and bad behavior.

I'd argue that `unregister()` should raise an exception if the reference count is non-zero.

Since services are not activated until they are needed, all we need to do is request that the services stop, then start again.

Can you try adding this code to ServiceManager, after `unregister()`?

```python3
    def restart(self, name: str):
        if name not in self.svcs:
            raise KeyError(f"Trying to restart unknown service {name!r}")

        svc = self.svcs[name]
        svc.stop()
        svc.await_stopped()
        svc.start()
        svc.await_ready()
```

Then call `app.svc.restart()` for each service name?

It's not the most elegant solution, but maybe you can test if that works? I can cook up something more fitting, if this works.

Reply:
If you rebase on top of `chrivers/service-improvements`, there's new logic to help you out :)

The `ServiceManager` can now be iterated over, to get a list of service names.

Also, the `Service` class now has a `.restart()` function.

So something like this:

```python
for name in app.svc:
    app.svc.svcs[name].restart()
```

I'm still not convinced this is the right way to do it, but let's see if it works? :)

Reply:
Implemented and seems to work better.

Reply:
✅